File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.cpp
class Solution {
public:
    bool bfs(unordered_map<int, vector<int>>& adj,
             unordered_map<int, bool>& visited, int source, int destination) {
        queue<int> q;
        q.push(source);
        visited[source] = true;

        if (visited[destination] == true)
            return true;

        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (auto& i : adj[u]) {
                if (visited[i] == false) {
                    visited[i] = true;
                    q.push(i);

                    if (visited[destination] == true)
                        return true;
                }
            }
        }
        return false;
    }
    bool validPath(int n, vector<vector<int>>& edges, int source,
                   int destination) {
        unordered_map<int, vector<int>> adj;
        for (int i = 0; i < edges.size(); i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        unordered_map<int, bool> visited;
        return bfs(adj, visited, source, destination);
    }
};

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.go
func dfs(adj [][]int, source int, destination int, visited []bool) bool {
    if source == destination {
        return true
    }

    visited[source] = true

    for _, v := range adj[source] {
        if !visited[v] && dfs(adj, v, destination, visited) {
            return true
        }
    }
    return false
}

func validPath(n int, edges [][]int, source int, destination int) bool {
    adj := make([][]int, n)
    
    for _, edge := range edges {
        u := edge[0]
        v := edge[1]

        adj[u] = append(adj[u], v)
        adj[v] = append(adj[v], u)
    }

    visited := make([]bool, n)

    path_exists := dfs(adj, source, destination, visited)

    return path_exists
}